util/conditions/experimental/summary.go

```diff
@@ -0,0 +1,129 @@
+	mergeStrategy                  MergeStrategy
+	conditionTypes                 []string
+	negativePolarityConditionTypes []string
+	stepCounter                    bool
+}
+
+// ApplyOptions applies the given list options on these options,
+// and then returns itself (for convenient chaining).
+func (o *SummaryOptions) ApplyOptions(opts []SummaryOption) *SummaryOptions {
+	for _, opt := range opts {
+		opt.ApplyToSummary(o)
+	}
+	return o
+}
+
+// NewSummaryCondition creates a new condition by summarizing a set of conditions from an object.
+// If any of the condition in scope does not exist in the source object, missing conditions are considered Unknown, reason NotYetReported.
+//
+// By default, the Aggregate condition has the same type of the source condition, but this can be changed by using
+// the OverrideType option.
+//
+// Additionally, it is possible to inject custom merge strategies using the WithMergeStrategy option or
+// to add a step counter to the generated message by using the WithStepCounter option.
+func NewSummaryCondition(obj runtime.Object, conditionType string, opts ...SummaryOption) (*metav1.Condition, error) {
+	summarizeOpt := &SummaryOptions{
+		mergeStrategy: newDefaultMergeStrategy(),
+	}
+	summarizeOpt.ApplyOptions(opts)
+
+	conditions, err := getConditionsWithOwnerInfo(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	expectedConditionTypes := sets.New[string](summarizeOpt.conditionTypes...)
+	existingConditionTypes := sets.New[string]()
+
+	// Drops all the conditions not in scope for the merge operation
+	conditionsInScope := make([]ConditionWithOwnerInfo, 0, len(expectedConditionTypes))
+	for _, condition := range conditions {
+		if !expectedConditionTypes.Has(condition.Type) {
+			continue
+		}
+		conditionsInScope = append(conditionsInScope, condition)
+		existingConditionTypes.Insert(condition.Type)
+	}
+
+	// Add the expected conditions which do net exists, so we are compliant with K8s guidelines
+	// (all missing conditions should be considered unknown).
+	// TODO: consider if we want to allow exception to this rule. e.g it is ok that HealthCheckSucceeded is missing.
+
+	diff := expectedConditionTypes.Difference(existingConditionTypes).UnsortedList()
+	if len(diff) > 0 {
+		conditionOwner := getConditionOwnerInfo(obj)
+
+		for _, conditionType := range diff {
+			conditionsInScope = append(conditionsInScope, ConditionWithOwnerInfo{
+				OwnerResource: conditionOwner,
+				Condition: metav1.Condition{
+					Type:    conditionType,
+					Status:  metav1.ConditionUnknown,
+					Reason:  NotYetReportedReason,
+					Message: fmt.Sprintf("Condition %s not yet reported", conditionType),
+				},
+			})
+		}
+	}
+
+	status, reason, message, err := summarizeOpt.mergeStrategy.Merge(
+		conditionsInScope,
+		summarizeOpt.conditionTypes,
+		sets.New[string](summarizeOpt.negativePolarityConditionTypes...),
+		summarizeOpt.stepCounter,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &metav1.Condition{
```

Comment:
Simila to other utils (we probably have to audit all utils), this condition has neither ObservedGeneration nor LastTransitionTime.

I think now I realized that not setting LastTransitionTime here is fine, because it is expected to be done by the Set func (if that is the case, let's please cleary document this in the function godocs).

For ObservedGeneration, I would have expected it to be set here

Reply:
There will only one point where we are setting ObservedGeneration, in Set.
All the other utils will rely on this (added notes to keep track of this)